csharp/src/Ice/Proxy.cs

```diff
@@ -322,76 +325,170 @@ public static class Proxy
+                            if (lastException != null)
+                            {
+                                throw ExceptionUtil.Throw(lastException);
+                            }
+                            else
+                            {
+                                Debug.Assert(response != null && response.ResultType == ResultType.Failure);
+                                return response;
                             }
                         }
-                        catch (OperationCanceledException)
+                        catch (ConnectionClosedByPeerException ex)
                         {
-                            throw; // Don't retry cancelled operations
+                            // Always retry a gracefully close connection. Don't exclude the connector after a
+                            // graceful close connection, in case it is the only connector available.
+                            lastException = ex;
+                            retryable = Retryable.AfterDelay;
+                        }
+                        catch (TransportException ex) when (request.IsIdempotent || !progressWrapper.IsSent)
+                        {
+                            // Retry transport exceptions if the request is idempotent or was not send
+                            lastException = ex;
+                            retryable = ex.Retryable;
+                        }
+
+                        if (response != null)
+                        {
+                            Debug.Assert(response.ResultType == ResultType.Failure);
+                            observer?.RemoteException();
+                            if (response.ReadSystemException(proxy.Communicator) is Exception systemException)
+                            {
+                                // 1.1 system exceptions
+                                if (systemException is ObjectNotExistException && reference.IsWellKnown)
+                                {
+                                    reference.LocatorInfo?.ClearCache(reference);
+                                    retryable = Retryable.AfterDelay;
+                                    afterDelay = TimeSpan.Zero;
+                                }
+                                else if (request.IsIdempotent && !progressWrapper.IsSent)
+                                {
+                                    retryable = Retryable.AfterDelay;
+                                    afterDelay = TimeSpan.Zero;
+                                }
+                            }
+                            else if (request.Encoding != Encoding.V20 &&
+                                     request.IsIdempotent &&
+                                     !progressWrapper.IsSent)
+                            {
+                                retryable = Retryable.AfterDelay;
+                                afterDelay = TimeSpan.Zero;
+                            }
```

Comment:
I'm not convinced we should keep retries for 1.1 exceptions that we no longer retry with 2.0... isn't it best to have consistent behavior on the client side whether or not you use a 1.1 or 2.0 server?

Reply:
Ok reworked this to keep consistent behavior with 1.1 and 2.0